src/Microsoft.AspNetCore.Razor.VSCode/src/Diagnostics/ReportIssuePanel.ts

```diff
@@ -0,0 +1,170 @@
+ * ------------------------------------------------------------------------------------------ */
+
+import * as clipboardy from 'clipboardy';
+import * as vscode from 'vscode';
+import { Trace } from 'vscode-jsonrpc';
+import { RazorLogger } from '../RazorLogger';
+import { ReportIssueCreator } from './ReportIssueCreator';
+import { ReportIssueDataCollector } from './ReportIssueDataCollector';
+import { ReportIssueDataCollectorFactory } from './ReportIssueDataCollectorFactory';
+
+export class ReportIssuePanel {
+    public static readonly viewType = 'razorReportIssue';
+
+    private panel: vscode.WebviewPanel | undefined;
+    private dataCollector: ReportIssueDataCollector | undefined;
+    private issueContent: string | undefined;
+
+    constructor(
+        private readonly dataCollectorFactory: ReportIssueDataCollectorFactory,
+        private readonly reportIssueCreator: ReportIssueCreator,
+        private readonly logger: RazorLogger) {
+    }
+
+    public async show() {
+        if (this.panel) {
+            this.panel.reveal(vscode.ViewColumn.Two);
+        } else {
+            this.panel = vscode.window.createWebviewPanel(
+                ReportIssuePanel.viewType,
+                'Report Razor Issue',
+                vscode.ViewColumn.Two, {
+                    enableScripts: true,
+                    // Dissallow any remote sources
+                    localResourceRoots: [],
+                });
+            this.attachToCurrentPanel();
+        }
+
+        await this.update();
+    }
+
+    public async revive(panel: vscode.WebviewPanel) {
+        this.panel = panel;
+        this.attachToCurrentPanel();
+        await this.update();
+    }
+
+    private attachToCurrentPanel() {
+        if (!this.panel) {
+            vscode.window.showErrorMessage('Unexpected error when attaching to report Razor issue window.');
+            return;
+        }
+
+        this.panel.webview.onDidReceiveMessage(async message => {
+            switch (message.command) {
+                case 'copyIssue':
+                    if (!this.issueContent) {
+                        if (!this.dataCollector) {
+                            vscode.window.showErrorMessage('You must first start the data collection before copying.');
+                            return;
+                        }
+                        const collectionResult = this.dataCollector.collect();
+                        this.issueContent = await this.reportIssueCreator.create(collectionResult);
+                        this.dataCollector = undefined;
+                    }
+
+                    await clipboardy.write(this.issueContent);
+                    vscode.window.showInformationMessage('Razor issue copied to clipboard');
+                    return;
+                case 'startIssue':
+                    this.issueContent = undefined;
+                    this.dataCollector = this.dataCollectorFactory.create();
```

Comment:
Should we also check if there was a previous `dataCollector` and if so, dispose it?

Otherwise, if the user clicks "start" repeatedly, they will be leaking `focusRegistration` and `logRegistration`.

As a side note, it's awkward that `start` is separate from the `ReportIssueDataCollector` constructor. It implies you could `start` multiple times, but if you do, that will leak as well. Would it be safer to make `start` implicit in the constructor?

Reply:
> As a side note, it's awkward that `start` is separate from the `ReportIssueDataCollector` constructor. It implies you could `start` multiple times, but if you do, that will leak as well. Would it be safer to make `start` implicit in the constructor?

hmm, true I could definitely do that. When I wrote the first iteration of this the flow was slightly different which warranted a separate start. Now that's not the case i'll fix.

> Should we also check if there was a previous `dataCollector` and if so, dispose it?
> 
> Otherwise, if the user clicks "start" repeatedly, they will be leaking `focusRegistration` and `logRegistration`.

Great catch, yes we should!